Created SequenceLexicoder to resolve #1252 (2.0)

reggert opened the pull request:
This provides a replacement for ListLexicoder that supports empty lists, resolving #1252 .

This PR replaces #1253 , which was originally targeted to `master` and Java 11. This one targets the `2.0` branch and Java 8.

milleruntime commented:
Thanks @reggert .  Just FYI you should be able to Edit a PR at the top and change the target branch by clicking the Edit button next to the title.

reggert commented:
> Thanks @reggert . Just FYI you should be able to Edit a PR at the top and change the target branch by clicking the Edit button next to the title.

Ah. I tried that, but I didn't notice that the target branch changed into a drop-down when I click "Edit". It looked like I could only edit the title.

reggert commented:
The previous PR was reviewed by @keith-turner . I've made a few tweaks since then (in addition to the rebase), including addressing his comments. As far as I know, I have no further changes to make.

reggert commented:
This is intended as a replacement for `ListLexicoder`. Delegating to it would seem to defeat that purpose. 
Additionally, prefixing the non-empty list encodings with a single byte would have a slight negative impact on the performance of lexicographical sorts, as well have undesirable effects on any rowids generated using it (affecting splits and shell output), as the overwhelming majority of encoded values (i.e., every non-empty list) would have the same first byte. Marking non-empty lists with a trailing byte, which this implementation does, avoids these (small but nonetheless annoying) issues.

In any case, I don't think delegating to `ListLexicoder` would have as much of a "simplifying" effect on the implementation as you think it does, since the byte would still need to be added before encoding, removed prior to decoding, with appropriate checks added to validate that nothing unexpected is in the data. For this implementation, I just started from the existing implementation of `ListLexicoder` and added the appropriate changes to handle empty lists, then applied a few optimizations, validations, and general cleanup to make it a bit more robust. I also modified the implementation of `ListLexicoder` to validate that it isn't applied to empty lists. Delegating to it would mean an unnecessary check would be taking place on data that was already validated. It would also necessitate an otherwise unneeded array copy after invoking `ListLexicoder.encode` to order to add the prefix/suffix byte, whereas this implementation avoids that.

reggert commented:
Standard formatting applied as requested.

ctubbsii commented:
> This is intended as a replacement for `ListLexicoder`. Delegating to it would seem to defeat that purpose.

No, they can still share implementation, where it makes sense to do so.

> Additionally, prefixing the non-empty list encodings with a single byte would have a slight negative impact...

Agreed. You're right. Putting the "empty/not-empty" marker at the end is better than at the beginning.

> In any case, I don't think delegating to ListLexicoder would have as much of a "simplifying" effect on the implementation...

I'm merely suggesting a slight refactor to use a common implementation to decode after checking the input array length when decoding, and to use a common implementation to encode prior to appending a marker when encoding. A common implementation should be possible, given that the only difference between the two would be the presence of an additional marker byte at the end (which is ignored, and present only to distinguish from the empty case where length == 0).

reggert commented:
Most of the heavy lifting is already being done by the `ByteUtils.concat` and `ByteUtils.split` methods, which are called by both implementations, effectively providing the common implementation. There isn't much else that is really shareable between the two classes, due to subtle differences in how they handle the ends of the lists, as well as the validation checks that they do, at least not without sacrificing efficiency. The `SequenceLexicoder.decodeUnchecked` method is only longer than the equivalent method in `ListLexicoder` because I added some sanity checks and split out some chained function calls that were all on one line in `ListLexicoder` into multiple lines in `SequenceLexicoder` for the sake of code clarity. I could have done the same for `ListLexicoder`, but didn't want to modify code that I wasn't altering the behavior of.

reggert commented:
On a side note, the automated "standard" formatting did some truly horrifying things to a few parts of the code, like splitting

```java
assert escapedElements.length > 0 : "ByteUtils.split always returns a minimum of 1 element, even for empty input";
```

into

```java
assert escapedElements.length
        > 0 : "ByteUtils.split always returns a minimum of 1 element, even for empty input";
```

Who inserts a line break into the middle of a numeric comparison? ;-)

ctubbsii commented:
> Most of the heavy lifting is already being done by the `ByteUtils.concat` and `ByteUtils.split` methods, which are called by both implementations, effectively providing the common implementation. There isn't much else that is really shareable between the two classes...

Okay, that's fair. If there's something we can think of later to deduplicate common code, it can be done later, as a follow-on.

> ...the automated "standard" formatting did some truly horrifying things...

Agreed. That kind of stuff has been bothering me a bit, too. I recently changed some of the formatting rules to support Java 11 and to wrap more aggressively in some cases, but the rules probably still need some more tweaking.

Our formatting rules are based on the Eclipse JDT formatter, used by the formatter-maven-plugin, and the config is located at: https://github.com/apache/accumulo/blob/master/contrib/Eclipse-Accumulo-Codestyle.xml ; feel free to open a PR to change the config or discuss on the mailing list any proposals you might have to improve the formatting. Just keep in mind that it's more important that we converge on a ruleset that can be automatically enforced, even if it's slightly inferior aesthetically, so changes should be within the capabilities of the JDT formatter.

reggert commented:
> The main differences between your implementation and my proposed one is that yours adds an extra delimiter on the end (via the concat method), even in the case of an empty list, whereas my suggestion wouldn't add one when the list is empty. But, either way is probably fine.
> 

No, it doesn't. An empty list is encoded as an empty byte array. There will only be a delimiter if there is at least one element. A single empty element will be encoded as a single delimiter byte.

The implementation of `ByteUtils.concat`  uses this block at the end of the field loop to insert delimiters:
```java
if (index < ret.length)
   ret[index++] = 0x00;
```
If `if` effectively prevents the loop from adding a delimiter to the end of the list, since the length of `ret` is determined by the expression `len + fields.length - 1` (where `len` is the sum of the encoded field lengths).

reggert commented on core/src/main/java/org/apache/accumulo/core/client/lexicoder/SequenceLexicoder.java:
```diff
@@ -0,0 +1,94 @@
+  @Override
+  protected List<E> decodeUnchecked(final byte[] b, final int offset, final int len) {
+    final byte[][] escapedElements = split(b, offset, len);
+    assert escapedElements.length
```
The assertion is checking (and documenting) a code assumption, not validating arguments. If assertions are disabled, that's fine, because this is only checking that something that should never happen doesn't happen.

reggert replied:
Also note that if the assumption is invalid and assertions are disabled, an `ArrayIndexOutOfBoundsException` will be thrown on the next line.

ctubbsii replied:
The decode method is effectively validating arbitrary user input (since it is the responsibility of the user's client code to ensure that what is passed to the decode method is something which was written after being emitted from the corresponding encode method).

Since this is validating arbitrary user input at a public API boundary, rather than asserting internal program correctness, an explicit check of the input is more appropriate than an assert statement here.

Also, `IllegalArgumentException` would be more appropriate than `ArrayIndexOutOfBoundsException` because the latter exposes, and is dependent upon, implementation details. `IllegalArgumentExcept`, on the other hand, is a helpful API exception for the caller that explicitly says there's a problem with the argument they provided.

keith-turner replied:
@reggert sorry for not getting back to this, it fell off my radar.

>  because this is only checking that something that should never happen doesn't happen.

Is this statement based on the assumption that users will only decode data that was encoded with this lexicoder?  Is the following use case possible?

  * User uses this lexicoder to encode data and store it in Accumulo column family.
  * User accidentally passes Accumulo row to this code for decoding and it triggers.  Could it throw the array out bounds exception in this case?

If the above is possible a user facing error message would be a nice touch to an already really nice change.  Maybe a message like `Either this data was corrupted or not encoded with "+this.getClass().getName()`.

reggert replied:
The assertion isn't checking user input. It's checking the return value of `ByteUtils.split`, and verifying that it behaves as expected.

keith-turner replied:
> The assertion isn't checking user input. It's checking the return value of ByteUtils.split, and verifying that it behaves as expected.

I see.  I misunderstood what that code was trying to do.

ctubbsii commented:
> No, it doesn't. An empty list is encoded as an empty byte array. There will only be a delimiter if there is at least one element. A single empty element will be encoded as a single delimiter byte.

Okay. I see that now. I guess I wasn't thinking clearly. Thanks for clarifying. :smiley_cat:

It becomes more clear to me when I write it out as a grammar:
*  `SEQUENCE → ( ESCAPED_ELEM DELIM )* EMPTY`
*  `LIST → ( ESCAPED_ELEM DELIM )* ESCAPED_ELEM`

keith-turner commented on core/src/main/java/org/apache/accumulo/core/client/lexicoder/SequenceLexicoder.java:
```diff
@@ -0,0 +1,94 @@
+ * in the elements escaped as two 0x1 bytes, and 0x1 bytes appearing in the elements escaped as 0x1
+ * and 0x2 bytes. The list is terminated with a final delimiter, with no bytes following it.
+ *
+ * @since 2.0.0
```
Glad to see this.

ctubbsii commented:
Thank you for your pull request, @reggert 
I've merged it into the 2.0 branch and it should release with 2.0.0.

If you'd like to be added to our [list of contributors on our website][1], feel free to open a pull request to [edit that file with your information][2].

[1]: https://accumulo.apache.org/people/#contributors
[2]: https://github.com/apache/accumulo-website/edit/master/pages/people.md
